### DISCUSSION TITLE
Imported data model not playing nicely with \equal from ifthen

### DISCUSSION THREAD

**Discussion by WoosterUK** [jasperhabicht/jsonparse#31] [Q&A] [answered]
I'm trying to use jsonparse to read a json file that is supposed to contain customisation options for a document and then select various items to include in the document, depending on the options set. My thinking was that this should work with ifthen (or xifthen), through the ifthenelse command.

However, I'm having a really hard time figuring out how the data are actually being stored and formatted so that I can use appropriate commands. For example, I can't work out even how to get it so that a string read from a JSON file is correctly computed to be equal to its value! (See the attached file for more details.) That's before I even get into doing boolean logic with imported booleans, or arithmetic with integers, both of which I would need to do.

Any suggestions? Or is there some further development needed first?

[test.zip](https://github.com/user-attachments/files/17524970/test.zip)

**Comment by jasperhabicht** [comment, marked as answer]
Thank you for your question, however I am a bit reluctant to open some .zip from an unknown source, but maybe you can create a minimal working example?

If you need to grab a string from a JSON file or JSON data string for further processing, you might want to use the `\JSONParseExpandableValue` command. I never tested this in combination with the `ifthenelse` package, but I would assume that you need to feed it an expandable command (which also might need to be expanded first).

In general, the reason that testing with the `\equal` command has its hurdles is twofold: 

1) First, as stated in my previous comment and also in the package manual, if you want to further use the value you grab from JSON data via the `jsonparse` package, you need to use `\JSONParseExpandableValue`. 

2) Secondly, if you use `\JSONParseExpandableValue`, you need to know that the result is a `str` with specific catcodes. A `str` (string) has catcode 12 for all chararctes except spaces, while a `tl` (token list) has catcode 11 for letters. I should probably make this clear in the package manual (see #29 ).

Therefore, you need to test against a `str`. You can make use of the `\tl_to_str:n` command (or create a user command via `\NewExpandableDocumentCommand`) to convert a `tl` to a `str`.

The following is possible and should return "True":

```tex
\documentclass{article}
\usepackage{jsonparse, ifthen}

\ExplSyntaxOn
\NewExpandableDocumentCommand { \tlToStr } { m } {
    \tl_to_str:n {#1} % or \detokenize{#1}
}
\ExplSyntaxOff

\begin{document}

\JSONParse{\jsondata}{ { "string" : "foobarbaz" } }

\ifthenelse{
    \equal{\JSONParseExpandableValue{\jsondata}{string}}{\tlToStr{foobarbaz}}
}{True}{False}

\end{document}
```

---

Another way to do this is to use `\JSONParseValue[store in={\jsonstring}, rescan]` to store the rescanned return value into a token list which you then can do the comparison with. "Rescanning" here means that the category codes are reset to those that TeX initially uses for each token in the token list. 

```tex
\documentclass{article}
\usepackage{jsonparse, ifthen}

\begin{document}

\JSONParse{\jsondata}{ { "string" : "foobarbaz" } }

\JSONParseValue[store in={\jsonstring}, rescan]{\jsondata}{string}

\ifthenelse{
    \equal{\jsonstring}{foobarbaz}
}{True}{False}

\end{document}
```

**Comment by WoosterUK** [comment]
Okay, that's fair on not downloading zip files. I've figured out how to get a demonstration tex and a json to you -- GitHub doesn't take tex files as attachments, rather frustratingly.

JSON: [test.json](https://github.com/user-attachments/files/17530587/test.json)
TEX: https://pastebin.com/yaJMmFJL

But it sounds like you had inferred correctly what I was wanting to do.

TeX isn't really my area in terms of coding expertise; I'm a modestly sensible LaTeX user but I'm not really in the low-level TeXnicalities of data types etc. I'll give your suggestion a go and see how that runs. Thanks!

**Comment by jasperhabicht** [comment]
Having checked your code, I can only repeat what I said above. 

Since `\tl_to_str:n` is essentially just `\detokenize`, you can define as follows:

```tex
\def\str{\JSONParseExpandableValue{\test}{str}}
\def\strA{\detokenize{string}}
```

Then, `\ifthenelse{\equal{\str}{\strA}}{true}{false}` will return "true".

You don't need to do this if comparing integers, since digits and punctuation marks have catcode 12 per default. 

Another way could be to rescan what is returned by `\JSONParseExpandableValue{\test}{str}`, but this is probably not expandable, so it won't directly work with `\ifthenelse`. 

Note that you can just write `\JSONParseFromFile{\test}{test.json}` without quotation marks around the file name. I am actually unsure why it also works with quotes ... 

I updated the documentation to hopefully clarify a bit better how the parsing works and which kind of return values the user can expect.

**Comment by WoosterUK** [comment]
Aha, thank you! The quotes came from my forgetting briefly that I was in LaTeX... I learnt LaTeX a long old time ago, and these days tend to do more work in Python.

`\detokenize` is great and a much slicker solution than whatever sorcery one is engaged in by invoking `\ExplSyntaxOn`. Not that I really still understand the internals of whatever is going on there, but hey; better things to worry about!
